ports/raspberrypi/common-hal/pulseio/PulseIn.c

```diff
@@ -0,0 +1,231 @@
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2017-2018 Scott Shawcroft for Adafruit Industries
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "src/rp2_common/hardware_gpio/include/hardware/gpio.h"
+#include "src/rp2_common/hardware_irq/include/hardware/irq.h"
+
+#include <stdint.h>
+
+#include "background.h"
+#include "mpconfigport.h"
+#include "py/gc.h"
+#include "py/runtime.h"
+#include "shared-bindings/microcontroller/__init__.h"
+#include "shared-bindings/pulseio/PulseIn.h"
+#include "shared-bindings/microcontroller/Pin.h"
+#include "supervisor/shared/translate.h"
+#include "bindings/rp2pio/StateMachine.h"
+#include "common-hal/pulseio/PulseIn.h"
+
+pulseio_pulsein_obj_t* save_self;
+
+#define NO_PIN 0xff
+
+const uint16_t pulsein_program[] = {
+//   set    pindirs, 0 ; For input
+    0xe080,
+//   wait   0 pin, 0   ; Wait for first low to start
+    0x2020,
+//   irq wait 0        ; set IRQ 0 and wait
+    0xc020,
+//   .bitloop
+//   in    pins, 1 [1] ; sample every 3 cycles (2 instructions, 1 delay)
+    0x4101,
+//   jmp bitloop
+    0x0003,
+};
+
+void common_hal_pulseio_pulsein_construct(pulseio_pulsein_obj_t* self,
+        const mcu_pin_obj_t* pin, uint16_t maxlen, bool idle_state) {
+
+    self->buffer = (uint16_t *) m_malloc(maxlen * sizeof(uint16_t), false);
+    if (self->buffer == NULL) {
+        mp_raise_msg_varg(&mp_type_MemoryError, translate("Failed to allocate RX buffer of %d bytes"), maxlen * sizeof(uint16_t));
+    }
+    self->pin = pin->number;
+    self->maxlen = maxlen;
+    self->idle_state = idle_state;
+    self->start = 0;
+    self->len = 0;
+    save_self = self;
+
+    // Set everything up.
+    rp2pio_statemachine_obj_t state_machine;
+
+    bool ok = rp2pio_statemachine_construct(&state_machine,
+        pulsein_program, sizeof(pulsein_program) / sizeof(pulsein_program[0]),
+        125000 * 4,
```

Comment:
The frequency needs to be lower so that we have fewer words to deal with in the interrupt handler. As noted in the previous comment I will likely need to adjust this.